dj/api/nodes.py

```diff
@@ -546,3 +546,22 @@ def node_similarity(
     node2_ast = parse(node2.current.query)  # type: ignore
     similarity = node1_ast.similarity_score(node2_ast)
     return JSONResponse(status_code=200, content={"similarity": similarity})
+
+
+@router.get("/nodes/{name}/downstream/")
+def downstream_nodes(
+    name: str, *, node_type: NodeType = None, session: Session = Depends(get_session)
+) -> List[NodeMetadata]:
+    """
+    List all nodes that are downstream from the given node, filterable by type.
+    """
+    node = get_node_by_name(session=session, name=name)
+    queue = collections.deque([node])
```

Comment:
I'm curious why you used a queue here instead of just a filter. Wouldn't a filter be clearer and not require the for-loop over the children?

Reply:
Do you mean a filter for all of the node's children in the NodeRelationship table? But we'd still need to process all of the its children's children, hence the queue. I may be missing something about that table though.

Reply:
Ah I see, I was under the impression that `current.children` returned all children, not just the immediate ones. It makes sense because looking at when we create a node, we only include immediate parents. This might be ok but I'm wondering if we should update the `extract_dependencies_from_node` function to return all upstream parents. That would then make this endpoint work with just a filter (cc: @CircArgs). Approving this either way so that you're not blocked. 👍

Reply:
@samredai `extract_dependencies_from_node` at one time would recurse till hitting sources if you specified a depth, but I ultimately just made it do the direct dependencies because it wasn't clear if we would want to pay the cost upfront of finding all the dependencies or just the immediate ones and if we did want all of them I figured the user could easily just call it over and over. Should be an easy add though if we think we can use it in a few places and want to just consolidate

Reply:
@CircArgs yeah, I actually think we shouldn't store links to all children of a node unless we determine that we really need it from a performance perspective (i.e., people are making the same call over and over again). We'll probably need some more metadata on the node relationships as well to indicate that it's not an immediate child.